Add Fedex REST API rate and tracking functionality, to co-exist with Fedex WSDL

xqiu opened the pull request:
### Description (*)
Add Fedex REST API rate and tracking functionality, to co-exist with Fedex WSDL.
The reason is that Fedex WSDL still functions.  But Fedex starts to require some accounts to use REST API (in my actual case), and stopped one of my accounts to use WSDL.

### Fixed Issues (if relevant)
1. Fixes OpenMage/magento-lts#3308

### Manual testing scenarios (*)
1. Use Fedex shipping method, get quotation from Fedex test server
2. Use Fedex shipping method, get quotation from Fedex production server (I've used the code in production)
3. Use Fedex shipping method, to track the shipment in openmage backend and frontend, when clicking the fedex shipment number 123456789012 for both Fedex test and productions server (I've used the code in production)

### Questions or comments
The change requires PHP 8.1 and above, as the package shipstream/fedex-rest-sdk requires.

xqiu committed:
Add Fedex REST API usage for ship rate and tracking

Fedex is currently in transition from WSDL to REST API.  This change allows configurable usage of WSDL or REST API.

Note: uses shipstream/fedex-rest-sdk, but currently it's not updated with fedex tracking API schema, so need to use a patch branch for it.

Also composer installing shipstream/fedex-rest-sdk requires PHP 8.1+, so I've to change php version as well for this commit.

xqiu committed:
Update composer.json shipstream/fedex-rest-sdk to ^1.1.2

xqiu committed:
Merge branch 'OpenMage:main' into main

xqiu committed:
Add @xqiu as a contributor

xqiu committed:
Merge branch 'main' of https://github.com/xqiu/magento-lts

xqiu committed:
Merge branch 'main' into main

sreichel commented:
Someone using FE>DEX to test it?

xqiu committed:
Add retry logic with shipDatestamp increment for Fedex REST API quote

xqiu committed:
Merge branch 'main' of https://github.com/xqiu/magento-lts

xqiu committed:
Merge branch 'main' into main

mattdavenport commented:
Figured I would try the Copilot reviewer here. FedEx will be retiring Web Services on 6/1/26, and as @xqiu mentioned, new accounts can no longer access the SOAP API. I'll keep an eye on this one and do some testing myself in the coming weeks, but if any other maintainer has an account (and the time 😄) to test it would be great to push this through.

xqiu commented on app/code/core/Mage/Usa/Model/Shipping/Carrier/Fedex.php:
```diff
@@ -1605,4 +1647,369 @@ public function getDeliveryConfirmationTypes(?Varien_Object $params = null)
+                new RequestedPackageLineItem(
+                    weight: new Weight(
+                        units: $this->getConfigData('unit_of_measure'),
+                        value: $this->getTotalNumOfBoxes($request->getPackageWeight()),
```
This is the same logic as the WebService code. getTotalNumOfBoxes will calculate the number of boxes requested, and get the weight of the one box.  Fedex code logic will calculate the total price based on the number of boxes calculated here later.

xqiu commented on app/code/core/Mage/Usa/Model/Shipping/Carrier/Fedex.php:
```diff
@@ -1605,4 +1647,369 @@ public function getDeliveryConfirmationTypes(?Varien_Object $params = null)
+            if($storeCurrencyCode != $currencyCode){    
+                $currencyStore = Mage::getModel('directory/currency')->load($storeCurrencyCode);
+                $currencyCurrent = Mage::getModel('directory/currency')->load($currencyCode);
+                $rate = round($rate / $currencyStore->getRate($currencyCurrent));
```
Gpt5 first insisted Copilot's opinion.  But Grok shows different.  When input gpt5 result to Grok, Grok insisted his opinion: ~~~
In Magento/OpenMage, $currencyStore->getRate($currencyCurrent) returns the exchange rate from the base currency ($currencyStore, e.g., CAD) to the FedEx currency ($currencyCurrent, e.g., USD), such as 0.74 USD per 1 CAD.
To convert $rate from USD to CAD, you divide by this rate: $rate / getRate(USD) (e.g., 100 / 0.74 ≈ 135.14 CAD).
ChatGPT’s suggestion to multiply assumes getRate() returns the rate from USD to CAD (e.g., 1.35 CAD per 1 USD), which is incorrect for Magento’s default behavior.
~~~
Gpt5 however, changed opinion after given grok's response.

My testing and production behavior shows original is correct.

xqiu committed:
Update based on copilot's review

Fixed spelling mistakes, typos and add a comment on why we use  $weight = $this->getTotalNumOfBoxes($request->getPackageWeight());

xqiu commented on app/code/core/Mage/Usa/etc/system.xml:
```diff
@@ -428,6 +428,35 @@
+                            <show_in_default>1</show_in_default>
+                            <show_in_website>1</show_in_website>
+                            <show_in_store>0</show_in_store>
+                            <comment>Fedex RestAPI need a seperate project for tracking, it's different with project that collect rate.</comment>
```
Fixed, pull request updated

xqiu commented on composer.json:
```diff
@@ -138,7 +139,7 @@
         },
         "platform": {
-            "php": "7.4"
+            "php": "8.1.28"
```
Fixed, pull request updated

xqiu committed:
Merge branch 'main' into main

xqiu committed:
Refactor duplicated lines from sonarqubecloud quality gate result

Refactor duplicated lines from sonarqubecloud quality gate result

xqiu committed:
Merge branch 'main' of https://github.com/xqiu/magento-lts

kiatng committed:
Merge branch 'main' into main

sreichel committed:
Merge branch 'main' into main

xqiu committed:
Update dependencies and updated composer.lock

Upgraded shipstream/fedex-rest-sdk to v1.2.0 and updated composer.lock

xqiu committed:
Improve FedEx carrier error handling and currency conversion from copilot code review

Refactored AccountNumber instantiation for compatibility. Enhanced retry logic with exponential backoff up to 32 seconds. Improved currency conversion by validating exchange rates before applying, skipping invalid rates. Fixed delivery date handling by checking for non-empty dateAndTimes array.

xqiu commented on app/code/core/Mage/Usa/Model/Shipping/Carrier/Fedex.php:
```diff
@@ -1599,4 +1653,353 @@ public function getDeliveryConfirmationTypes(?Varien_Object $params = null)
+        $this->_request = $request;
+
+        // Step 1: Create an AccountNumber object
+        $accountNumber = new AccountNumber(value: $this->getConfigData('account'));
```
resolved in checkin c6f23e5fcd79849b62286b27aae3aef5c95b1c8b

xqiu commented on app/code/core/Mage/Usa/Model/Shipping/Carrier/Fedex.php:
```diff
@@ -1599,4 +1653,353 @@ public function getDeliveryConfirmationTypes(?Varien_Object $params = null)
+                    $this->_result->append($error);
+                } else {
+                    // Optional: wait before retrying
+                    sleep(1);
```
resolved in checkin c6f23e5fcd79849b62286b27aae3aef5c95b1c8b

addison74 committed:
Merge branch 'main' into main

kiatng committed:
Merge branch 'main' into main

sreichel commented:
Closing here. Thanks.
